eva-accession-release-automation/gather_clustering_counts/gather_release_counts.py

```diff
@@ -0,0 +1,300 @@
+import argparse
+import csv
+import glob
+import os
+from collections import defaultdict, Counter
+from functools import lru_cache
+
+from ebi_eva_common_pyutils.command_utils import run_command_with_output
+from ebi_eva_common_pyutils.common_utils import pretty_print
+from ebi_eva_common_pyutils.logger import logging_config, AppLogger
+from ebi_eva_common_pyutils.metadata_utils import get_metadata_connection_handle
+from ebi_eva_common_pyutils.pg_utils import get_all_results_for_query
+
+
+logger = logging_config.get_logger(__name__)
+
+
+shell_script_dir = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'bash')
+
+assembly_table_name = 'eva_stats.release_rs_statistics_per_assembly'
+
+def find_link(key_set, dict1, dict2, source_linked_set1=None, source_linked_set2=None):
+    """
+    Assuming 2 dictionaries providing respectively the list of values linked to a key, and a list of keys linked to a
+    value, (in our case this is a list of assemblies linked to a taxonomy and the list of taxonomy linked to a assembly)
+    , this recursive function starts from one of the value and find all the related keys and values and provided them
+    in 2 frozen sets. For any key that belong to a relationship, this should provide the same pair of frozensets
+    regardless of the starting key.
+    """
+    if source_linked_set1 is None:
+        source_linked_set1 = set()
+    if source_linked_set2 is None:
+        source_linked_set2 = set()
+    linked_set1 = source_linked_set1.copy()
+    linked_set2 = source_linked_set2.copy()
+    for key1 in key_set:
+        if key1 in dict1:
+            for value1 in dict1.get(key1):
+                linked_set2.add(value1)
+                if value1 in dict2:
+                    linked_set1.update(dict2.get(value1))
+    # if one of the set is still growing we check again
+    if linked_set1 != source_linked_set1 or linked_set2 != source_linked_set2:
+        tmp_linked_set1, tmp_linked_set2 = find_link(linked_set1, dict1, dict2, linked_set1, linked_set2)
```

Comment:
I'm interpreting this algorithm (in computer science-speak) as finding connected components in a bipartite graph, if so that's pretty cool!

Not sure if efficiency is an issue, but if you maintained a list of "visited" vertices you could ensure that you don't repeat work in the recursion (as I think `linked_set1` is always a superset of `key_set`).